okio/src/main/java/okio/HashingSink.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+import static okio.Util.checkOffsetAndCount;
+
+/**
+ * A sink that computes a hash of the full stream of bytes it has accepted. To use, create an
+ * instance with your preferred hash algorithm. Write all of the data to the sink and then call
+ * {@link #hash()} to compute the final hash value.
+ *
+ * <p>In this example we use {@code HashingSink} with a {@link BufferedSink} to make writing to the
+ * sink easier. <pre>   {@code
+ *
+ *   HashingSink hashingSink = HashingSink.sha256(s);
+ *   BufferedSink bufferedSink = Okio.buffer(hashingSink);
+ *
+ *   ... // Write to bufferedSink and either flush or close it.
+ *
+ *   ByteString hash = hashingSink.hash();
+ * }</pre>
+ */
+public final class HashingSink extends ForwardingSink {
+  private final MessageDigest messageDigest;
+
+  /** Returns a sink that uses the obsolete MD5 hash algorithm. */
+  public static HashingSink md5(Sink sink) {
+    try {
+      return new HashingSink(sink, MessageDigest.getInstance("MD5"));
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  /** Returns a sink that uses the obsolete SHA-1 hash algorithm. */
+  public static HashingSink sha1(Sink sink) {
+    try {
+      return new HashingSink(sink, MessageDigest.getInstance("SHA-1"));
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  /** Returns a sink that uses the SHA-256 hash algorithm. */
+  public static HashingSink sha256(Sink sink) {
+    try {
+      return new HashingSink(sink, MessageDigest.getInstance("SHA-256"));
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  private HashingSink(Sink sink, MessageDigest messageDigest) {
```

Comment:
None. In theory we later add hashes not-exposed by MessageDigest like MurmurHash3, so I don’t want to couple this to that API.